### ISSUE TITLE
Mapping of Imports and modifiers on JDT

### ISSUE THREAD

**Issue by martinezmatias** [GumTreeDiff/gumtree#212]
Hi @jrfaller, @monperrus 

I open this issue to discuss a particular issue that we found and discussed when we did the manual analysis for the DAT paper.
The issue was related to the generation of incorrect add/remove of imports and/or modifiers from the Classic.

The following image shows one of those cases: 
<img width="1319" alt="Screen Shot 2021-01-26 at 17 59 26" src="https://user-images.githubusercontent.com/7804830/105959458-31a46280-607c-11eb-9307-747c6f4d976a.png">

This issue summarizes the finding of investigating those strange outputs.

To study the issue, I created different test cases which are pushed in my folk:
https://github.com/martinezmatias/gumtree/blob/jdt_mapping/gen.jdt/src/test/java/com/github/gumtreediff/gen/jdt/TestJdtMatching.java

There are failing test cases that expose the unexpected output (i.e., that contain incorrect remove/add actions) and passing test cases that execute, for the same cases, GumTree with different configuration that produces the expected output.

I studied two particular cases.

First,  incorrect add/remove actions of Import declaration.
I wrote a test that exposes this case: https://github.com/martinezmatias/gumtree/blob/jdt_mapping/gen.jdt/src/test/java/com/github/gumtreediff/gen/jdt/TestJdtMatching.java#L234

The difference between the two files is the update of one constant. However, the output of Classic is a larger script with addition/remove of import declarations
Those are caused by the default value of *min height* used by the  SubTreeMatcher: In GT3.0 the default value is *2* , however, in the version 2, the default value is 1.
As the height of Import declaration node is 1, then, the SubTreeMarcher (e.g., Greedy used by Classic) does not map the imports.
If we configure the min value as 1, it works fine as shows the test: https://github.com/martinezmatias/gumtree/blob/jdt_mapping/gen.jdt/src/test/java/com/github/gumtreediff/gen/jdt/TestJdtMatching.java#L372

Now, the question is: as the height of import node is always the same on JDT, why sometimes we have that incorrect output with add/remove imports and  why sometimes the output is correct i.e. imports are correctly mapped? 
The answer I found is: The difference on the output is thus given by the second stage of GumTree. In particular, for Classic is the Greedy BottomUp Matcher and its hyper parameter *size* of the trees.
If the sizes of the trees  are smaller that the size threshold, then BottomUp matcher tries to match the unmapped elements via the last chance: https://github.com/martinezmatias/gumtree/blob/develop/core/src/main/java/com/github/gumtreediff/matchers/heuristic/gt/AbstractBottomUpMatcher.java#L80
Otherwise, size of trees larger than threshold, those unmapped nodes (including the imports) remain unmapped.
On one test case I play with the *size* hyper parameter value to mapping or unmapping the import nodes: https://github.com/martinezmatias/gumtree/blob/jdt_mapping/gen.jdt/src/test/java/com/github/gumtreediff/gen/jdt/TestJdtMatching.java#L351

The second case I studied  focuses on unmapping of modifiers.
The following test case exposes that case:  https://github.com/martinezmatias/gumtree/blob/jdt_mapping/gen.jdt/src/test/java/com/github/gumtreediff/gen/jdt/TestJdtMatching.java#L56

Here, the *modifiers*  nodes are never mapped by the SubTreeMarcher  as their height is always zero.
Thus, their mapping (or not) will depend on the size of the root tree given by hyperparameter from BottomUpMatcher.
In the  following test I change the default size to produce the expected output: https://github.com/martinezmatias/gumtree/blob/jdt_mapping/gen.jdt/src/test/java/com/github/gumtreediff/gen/jdt/TestJdtMatching.java#L163

To sum up: I wanted to report the findings after inspecting those cases.
I don't see any changes on GumTree, except the potential change of  *min height*  from 2 to 1 (as is used in GumTree2).

Some of the test cases I created are failing because those expect the correct behaviour.
I can modify those tests to produce only  passing test, and then we can integrate to the GumTree main repo via a PR.
WDYT?

Cheers
Matias

**Comment by jrfaller** [comment]
Hi @martinezmatias and thanks a lot for the thorough feedback!

You are exactly right and I was aware of this particular frequent problem. There is not much we can do as you remark with the classic gumtree bottom up matcher. However, simple gumtree bottom up matcher, as it do not have a threshold for last chance matches, usually work well for these cases.

**Comment by jrfaller** [comment]
BTW, you can look into the interesting discussion here in issue #208 with an interesting solution ;)

**Comment by jrfaller** [comment]
The idea of #208 has been prototyped, you can try it out ;)

Cheers.
